Source/Core/VideoCommon/UberShaderPixel.cpp

```diff
@@ -0,0 +1,716 @@
+		"			// Texture is enabled\n"
+		"			uint sampler_num = %s;\n", BitfieldExtract("order", TwoTevStageOrders().texmap0).c_str());
+	out.Write(
+		"			uint tex_coord = %s;\n", BitfieldExtract("order", TwoTevStageOrders().texcoord0).c_str());
+	out.Write(
+		"\n"
+		"			// TODO: there is an optional perspective divide here (not to mention all of indirect)\n"
+		"			int2 fixedPoint_uv = itrunc(tex[tex_coord].xy * " I_TEXDIMS"[sampler_num].zw * 128.0);\n"
+		"			float2 uv = (float2(fixedPoint_uv) / 128.0) * " I_TEXDIMS"[sampler_num].xy;\n"
+		"\n"
+		"			int4 color = sampleTexture(sampler_num, uv);\n"
+		"\n"
+		"			uint swap = %s;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.tswap).c_str());
+	out.Write(
+		//"			s.TexColor.r = Swap(0, color);\n");
+		"			s.TexColor.r = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2u]", TevKSel().swap1).c_str());
+	out.Write(
+		//"			s.TexColor.g = Swap(1, color);\n");
+		"			s.TexColor.g = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2u]", TevKSel().swap2).c_str());
+	out.Write(
+		//"			swap = swap + 1;\n"
+		//"			s.TexColor.b = Swap(2, color);\n");
+		"			s.TexColor.b = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2u+1u]", TevKSel().swap1).c_str());
+	out.Write(
+		//"			s.TexColor.a = Swap(3, color);\n");
+		"			s.TexColor.a = Swap(3u-%s, color);\n", BitfieldExtract("bpmem_tevksel[swap*2u+1u]", TevKSel().swap2).c_str());
+	out.Write(
+		"		} else {\n"
+		"			// Texture is disabled\n"
+		"			s.TexColor = int4(255, 255, 255, 255);\n"
+		"		}\n"
+		"\n");
+	}
+
+	out.Write(
+		"		// Select Konst for stage\n"
+		"		// TODO: a switch case might be better here than an dynamically indexed uniform lookup\n"
+		"		uint tevksel = bpmem_tevksel[stage>>1];\n"
+		"		if ((stage & 1u) == 0u)\n"
+		"			s.KonstColor = int4(konstLookup[%s].rgb, konstLookup[%s].a);\n",
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kcsel0).c_str(),
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kasel0).c_str());
+	out.Write(
+		"		else\n"
+		"			s.KonstColor = int4(konstLookup[%s].rgb, konstLookup[%s].a);\n\n",
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kcsel1).c_str(),
+		BitfieldExtract("tevksel", bpmem.tevksel[0].kasel1).c_str());
+	out.Write(
+		"\n");
+
+	out.Write(
+		"		// Select Ras for stage\n"
+		"		uint ras = %s;\n", BitfieldExtract("order", TwoTevStageOrders().colorchan0).c_str());
+	out.Write(
+		"		if (ras < 2u) { // Lighting Channel 0 or 1\n"
+		"			int4 color = (ras == 0u) ? icolors_0 : icolors_1;\n"
+		"			uint swap = %s << 1;\n", BitfieldExtract("ac", TevStageCombiner().alphaC.rswap).c_str());
+	out.Write(
+		//"			s.RasColor.r = Swap(0, color);\n");
+		"			s.RasColor.r = Swap(%s, color);\n", BitfieldExtract("bpmem_tevksel[swap]", TevKSel().swap1).c_str());
+	out.Write(
+		//"			s.RasColor.g = Swap(1, color);\n");
+		"			s.RasColor.g = Swap(%s, color);\n", BitfieldExtract("bpmem_tevksel[swap]", TevKSel().swap2).c_str());
+	out.Write(
+		"			swap = swap + 1u;\n"
```

Comment:
I guess this formulation for swap is correct, but could you please use the same way to calculate it for both texture and ras?